[TOSA] Handle float<->bool cast via i8 in tosaCastTensorToType

Lallapallooza opened the pull request:
Add intermediate i8 cast to support float<->bool conversions, which TOSA doesn’t allow directly. Fixes legalization failures for such cases.

sjarus reviewed (commented):
What legalization failures are now fixed ? Does projects/pt1/e2e_testing/xfail_sets.py need to be updated ?

Lallapallooza commented:
> What legalization failures are now fixed ? Does projects/pt1/e2e_testing/xfail_sets.py need to be updated ?

I have added dedicated tests to cover the affected legalization flow. While the existing e2e test TypeConversionI1ToF32Module_basic was already passing before, the generated TOSA IR was not standards-compliant and failed verification under strict profiles.

IR Before
```
module {
  func.func @main(%arg0: tensor<1x3x1x5xf32>) -> tensor<1x3x1x5xf32> {
    %0 = tosa.cast %arg0 : (tensor<1x3x1x5xf32>) -> tensor<1x3x1x5xi1>
    %1 = tosa.cast %0 : (tensor<1x3x1x5xi1>) -> tensor<1x3x1x5xf32>
    return %1 : tensor<1x3x1x5xf32>
  }
}
```
Which is failed to verify
```
$ ./build/bin/mlir-opt --tosa-validate="profile=pro_int,pro_fp strict-op-spec-alignment" tosa.mlir
tosa.mlir:3:10: error: 'tosa.cast' op illegal: operation operand/result data types did not align with any profile or extension, got (f32,i1), did you mean (i8,i1)? Otherwise, please refer to the 'supported data types' for 'tosa.cast' in the specification.
    %0 = tosa.cast %arg0 : (tensor<1x3x1x5xf32>) -> tensor<1x3x1x5xi1>
         ^
tosa.mlir:3:10: note: see current operation: %0 = "tosa.cast"(%arg0) : (tensor<1x3x1x5xf32>) -> tensor<1x3x1x5xi1>
tosa.mlir:4:10: error: 'tosa.cast' op illegal: operation operand/result data types did not align with any profile or extension, got (i1,f32), did you mean (i1,i8)? Otherwise, please refer to the 'supported data types' for 'tosa.cast' in the specification.
    %1 = tosa.cast %0 : (tensor<1x3x1x5xi1>) -> tensor<1x3x1x5xf32>
         ^
tosa.mlir:4:10: note: see current operation: %1 = "tosa.cast"(%0) : (tensor<1x3x1x5xi1>) -> tensor<1x3x1x5xf32> 
```

With new patch the generated TOSA is
```
module {
  func.func @main(%arg0: tensor<1x3x1x5xf32>) -> tensor<1x3x1x5xf32> {
    %0 = "tosa.const"() <{values = dense<0.000000e+00> : tensor<1x1x1x1xf32>}> : () -> tensor<1x1x1x1xf32>
    %1 = tosa.equal %arg0, %0 : (tensor<1x3x1x5xf32>, tensor<1x1x1x1xf32>) -> tensor<1x3x1x5xi1>
    %2 = tosa.logical_not %1 : (tensor<1x3x1x5xi1>) -> tensor<1x3x1x5xi1>
    %3 = tosa.cast %2 : (tensor<1x3x1x5xi1>) -> tensor<1x3x1x5xi8>
    %4 = tosa.cast %3 : (tensor<1x3x1x5xi8>) -> tensor<1x3x1x5xf32>
    return %4 : tensor<1x3x1x5xf32>
  }
}
```
This passes the validator and produces results identical to PyTorch.

sjarus reviewed (approved):
Nice addition, thanks!
